src/main/java/org/openlmis/integration/dhis2/service/ReferenceDataService.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * This program is part of the OpenLMIS logistics management information system platform software.
+ * Copyright © 2017 VillageReach
+ *
+ * This program is free software: you can redistribute it and/or modify it under the terms
+ * of the GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details. You should have received a copy of
+ * the GNU Affero General Public License along with this program. If not, see
+ * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
+ */
+
+package org.openlmis.integration.dhis2.service;
+
+import static org.openlmis.integration.dhis2.util.RequestHelper.createEntity;
+
+import java.net.URI;
+import org.openlmis.integration.dhis2.dto.referencedata.MinimalFacilityDto;
+import org.openlmis.integration.dhis2.dto.referencedata.OrderableDto;
+import org.openlmis.integration.dhis2.dto.referencedata.PageDto;
+import org.openlmis.integration.dhis2.exception.ResponseParsingException;
+import org.openlmis.integration.dhis2.exception.RestOperationException;
+import org.openlmis.integration.dhis2.i18n.MessageKeys;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+@Service
+public class ReferenceDataService {
+
+  private static final String API_URL = "/api/";
+  private static final String FACILITIES_RESOUCE_PATH = "facilities/";
+  private static final String ORDERABLES_RESOUCE_PATH = "orderables/";
+
+  @Value("${service.url}")
+  private String serviceUrl;
+
+  @Autowired
+  private RestTemplate restTemplate;
+
+  @Autowired
+  private ReferenceDataAuthService authService;
+
+  /**
+   * Retrieves {@link PageDto} of {@link MinimalFacilityDto} objects from referencedata service.
+   *
+   * @return page of MinimalFacilityDto objects.
+   */
+  public PageDto<MinimalFacilityDto> finaAllFacilities() {
+
+    try {
```

Comment:
I think 60-79 should be possible to extract to seperate method, I imagine something like: 
doResourceRequest<MinimalFacilityDto>(FACILITIES_RESOUCE_PATH, MinimalFacilityDto.class)

Because this is mostly the same as #finaAllOrderables

Reply:
I made changes